### PR TITLE
Version 1.0.1

### PR DESCRIPTION
**Pull request by cosminbosutar**
Version of that replaces the POST form with a codified JSON, solves security issues and has subscription support integrated.

### CODE REVIEW

File: upload/system/helper/Twispay_Status_Updater.php [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+            if ($order_recurring) {
+                Twispay_Status_Updater::updateSubscription($order_recurring, $decrypted, $that);
+            }
+        }
+
+        /**
+         * Update the status of an subscription according to the received server status.
+         *
+         * @param string order_id: The ID of the order to be updated.
+         * @param object decrypted: Decrypted order message.
```
Comment (dragos-ursu) [inline upload/system/helper/Twispay_Status_Updater.php]: The $decrypted parameter seem to always be sent as an array and not an object.

File: upload/catalog/model/extension/payment/twispay_transaction.php [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+     *
+     * @param array([key => value]) data - Array of data to be populated
+     *
+     * @return string - The query that was called | FALSE
+     *
+     */
+    public function insertTransaction($data)
+    {
+        $this->load->helper('Twispay_Status_Updater');
+        $data =json_decode(json_encode($data), TRUE);
```
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_transaction.php]: Is there a reason for this? In the code there are only calls that pass an array as the $data parameter.

File: upload/catalog/model/extension/payment/twispay_transaction.php [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        if (!empty($data['identifier'])) {
+            $data['identifier'] = (int)str_replace('_', '', $data['identifier']);
+        }
+
+        /** Construct the query based on $data object fields filtered by $colums keys */
+        $query = "INSERT INTO `" . DB_PREFIX . "twispay_transactions` SET ";
+        foreach ($data as $key => $value) {
+            if (!in_array($key, $columns)) {
+                unset($data[$key]);
```
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_transaction.php]: There is no further usage of the $data variable that would make use of unsetting a key at this point.

File: upload/catalog/model/extension/payment/twispay_transaction.php [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    /**
+     * Function that insert a record or update it if already exists.
+     *
+     * @param array([key => value]) data - Array of data to be populated
+     * @param boolean overwrite          - Allow a record to be updated or not
+     *
+     * @return {array|string} - Operation response
+     *
+     */
+    public function logTransaction($data, $overwrite = FALSE)
```
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_transaction.php]: Can't find any usage of this method, it is needed?

File: upload/catalog/model/extension/payment/twispay_transaction.php [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+     *
+     */
+    public function refund($trans_id, $order_id)
+    {
+        $this->load->helper('Twispay_Logger');
+        $this->load->helper('Twispay_Status_Updater');
+        $this->load->model('extension/payment/twispay_recurring');
+
+        $order_recurring = $this->model_extension_payment_twispay_recurring->getRecurringByOrderId($order_id);
+        $order_recurring_id = $order_recurring['order_recurring_id'];
```
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_transaction.php]: Can't find any usage of $order_recurring_id, it is needed?

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 3 above)
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_transaction.php]: There is no reason to unset $data[$key] at this point.

File: upload/catalog/model/extension/payment/twispay_recurring.php [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+     *
+     * @param array([key => value]) data - Array of data to be populated
+     * @param boolean overwrite          - Allow a record to be updated or not
+     *
+     * @return {array|string} - Operation response | FALSE
+     *
+     */
+    public function addRecurringTransaction($data, $overwrite = FALSE)
+    {
+        $data = json_decode(json_encode($data), TRUE);
```
Comment (dragos-ursu) [inline upload/catalog/model/extension/payment/twispay_recurring.php]: Any reason to have this here? Seem like $data is always sent as an array.

**Review by dragos-ursu — CHANGES_REQUESTED** [review]
Please see the comments.

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 2 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_transaction.php]: There was a case when $data was an object so json_decode(json_encode($data), TRUE); was converting it to an associative array. But its not needed anymore. I've removed it.

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 3 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_transaction.php]: Fixed.

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 4 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_transaction.php]: Is not needed anymode. I've removed it.

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 5 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_transaction.php]: Removed.

File: upload/catalog/model/extension/payment/twispay_transaction.php (hunk 3 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_transaction.php]: Fixed.

File: upload/catalog/model/extension/payment/twispay_recurring.php (hunk 6 above)
Comment (MichaelRotaru) [reply upload/catalog/model/extension/payment/twispay_recurring.php]: Fixed.

File: upload/system/helper/Twispay_Status_Updater.php (hunk 1 above)
Comment (MichaelRotaru) [reply upload/system/helper/Twispay_Status_Updater.php]: Fixed.

**Pull request merged**